Add support for Blocking and Non-blocking put-with-signal

naveen-rn opened the pull request:
### Proposal went mutliple changes in semantics and documentation after the initial reading - [refer to the following comment](https://github.com/openshmem-org/specification/pull/275#issuecomment-524505062) for the most recent update
Previously, we proposed the support for put-with-signal operation in #206. We worked on separate blocking and non-blocking proposals in #218 and #244. Based on multiple review comments and discussions, we have created a consolidated proposal which includes the following:

1. Support for blocking put-with-signal routines
2. Support for non-blocking put-with-signal routines
3. Library constants to select the type of signal update operation (SET and ADD)
4. Update to point-to-point synchronization routines to return signal value
5. Library constants to provide hints to the implementations on the type of update to be expected on the wait operation

In this proposal, apart from adding support for put-with-signal routines, we have update the point-to-point sync routines - especially ```shmem_wait_until``` to return the values from ```ivar``` on which the wait operation is performed. This would enable us to read the signal update efficiently using the wait operation. The need for the update on the wait operation are provided in the attached
[presentation.](https://github.com/openshmem-org/specification/files/3143910/reading-signal-put-with-signal.pptx)

In general, the proposed put-with-signal has the following atomicity guarantees. 

| signal update type | ```SHMEM_SIGNAL_SET``` | ```SHMEM_SIGNAL_ADD``` | ```shmem_atomics``` | p2p-sync |
|--------------------|-----------------------|------------------------|----------------------|-----------|
| ```SHMEM_SIGNAL_SET``` | exclusive | undefined | undefined | exclusive |
| ```SHMEM_SIGNAL_ADD``` | undefined | exclusive | undefined | exclusive |

The only possible way to read the signal in portable fashion is through the updated ```shmem_wait_until``` operation and passing the hints with the comparison operator would enable us to optimize the wait operation, if provided by the implementation.

Closes #206

manjugv commented on content/shmem_put_signal.tex:
```diff
@@ -0,0 +1,115 @@
+    performed on the remote \VAR{sig\_addr} signal data object. The completion
+    of signal update based on the \VAR{sig\_op} signal operator using the
+    \VAR{signal} flag on the remote \ac{PE} indicates the delivery of its
+    corresponding \dest{} data words into the data object on the remote \ac{PE}.
```
May F2F: Add a phrase about sig addr completion.

naveen-rn replied:
fixed

anshumang commented on content/shmem_put_signal_nbi.tex:
```diff
@@ -0,0 +1,103 @@
+void @\FuncDecl{shmem\_ctx\_putmem\_signal\_nbi}@(shmem_ctx_t ctx, void *dest, const void *source, size_t nelems, uint64_t *sig_addr, uint64_t signal, int sig_op, int pe);
+\end{CsynopsisCol}
+
+\begin{apiarguments}
```
Sharing a thought, does it make sense to not repeat description text for arguments for the same API (the `_nbi` and the other variant)?

naveen-rn replied:
Not sure if I understand the comment correctly - are you saying that the argument description can be combined with the blocking variant? If so, I think that would make the proposal not consistent with the other parts of the text. I'll close the comment for now and fix it in a separate cleanup PR is needed.

anshumang commented on content/shmem_put_signal_nbi.tex:
```diff
@@ -0,0 +1,103 @@
+    on the local \ac{PE} and delivered into the \dest{} array on the destination
+    \ac{PE}.
+
+    The delivery of \VAR{signal} flag on the remote \ac{PE} indicates only the
```
This is the only paragraph with text that provides additional info. Wondering if the repitition could be deleted in some way.

naveen-rn replied:
This comment is based on old text - things have moved around for the official reading. Closing this comment for now and let the author reopen if this is still an issue.

BryantLam commented:
Three options discussed regarding signal operation and interaction with `shmem_wait_until`.
1. Change `SHMEM_SIGNAL_UPDATE` from a hint to a requirement.
2. Introduce new API variants for `wait_until` that work specifically only for the signal operation to preserve performance.
3. Make signal operation atomic compatible with other atomic operations.
Option 2 was selected in a straw poll.

nspark commented on main_spec.tex:
```diff
@@ -23,7 +23,7 @@ \section{Execution Model}\label{subsec:execution_model}
 
 \section{Language Bindings and Conformance}\label{subsec:bindings}
-\input{content/language_bindings_and_conformance}
+\input{content/language_bindings_and_conformance}re
```
This seems like a stray typo.

naveen-rn replied:
fixed

nspark commented on content/shmem_signal_wait_until.tex:
```diff
@@ -0,0 +1,57 @@
+
+\begin{apiarguments}
+
+\apiargument{OUT}{sig\_addr}{A pointer to a remotely accessible variable.}
```
7/22 reading: "IN" not "OUT" argument

naveen-rn replied:
Fixed

anshumang commented:
Adding a point-to-point synchronization example that use `shmem_put_signal_nbi` would be useful given that this is the only non-blocking API that can be used for point-to-point synchronization.
Use of this API for point-to-point synchronization requires `shmem_quiet` at the initiator that none of the other synchronization APIs so.

naveen-rn commented:
Discussed with @anshuman offline and decided to add this example as a separate ticket once the main ticket is merged.

naveen-rn commented on content/p2p_sync_intro.tex:
```diff
@@ -3,8 +3,8 @@
 that memory access operations observe remote updates in the order enforced by
-the initiator \ac{PE} using the \FUNC{shmem\_fence} and \FUNC{shmem\_quiet}
-routines.
+the initiator \ac{PE} using the put-with-signal, \FUNC{shmem\_fence} and
```
Looks like a new PR is required - not needed as part of this proposal

naveen-rn commented:
## Update for voting in OpenSHMEM Sep 2019 F2F meeting:

### Proposal Overview
In this propoal, we are adding support for signaling operation in OpenSHMEM along with options to read the signal effectively. As part of this proposal, we introduce three signaling operations and a p2p signaling wait operation. We also introduced library constants ```SHMEM_SIGNAL_SET``` and ```SHMEM_SIGNAL_ADD`` to determine the type of signal update type.

#### Signaling Operations
1. blocking put-with-signal (```shmem_signal_put```), 
2. non-blocking put-with-signal (```shmem_signal_put_nbi```), and
3. signaling fetch (```shmem_signal_fetch```).
#### Signaling wait operation
1. ```shmem_signal_wait_until```
#### Atomicity
In general, the proposed put-with-signal has the following atomicity guarantees. 
1. exclusive with itself and with other signaling operation of the same signal update type,
2. exclusive with respect to all p2p sync operation, and
3. undefined with other OpenSHMEM AMO operations.
This semantic allows the most efficient implementation of the signaling operation and extract hardware optimizations.

| signal update type | ```SHMEM_SIGNAL_SET``` | ```SHMEM_SIGNAL_ADD``` | ```shmem_atomics``` | p2p-sync |
|--------------------|-----------------------|------------------------|----------------------|-----------|
| ```SHMEM_SIGNAL_SET``` | exclusive | undefined | undefined | exclusive |
| ```SHMEM_SIGNAL_ADD``` | undefined | exclusive | undefined | exclusive |
#### Reading Signal Update
Because of the above semantics, signal updates can be read only through special routines : ```shmem_signal_fetch``` and ```shmem_signal_wait_until```. While ```shmem_signal_fetch``` is introduced to exclusively read the signal update, ```shmem_signal_wait_until``` can be used to wait on a signal update and return the updated signal value.

### History
We introduced blocking put-with-signal and non-blocking put-with-signal as separate APIs in #218 and #244. In this proposal, we have consolidated those two proposals and incorporated all the suggested changes from the OpenSHMEM specification committee.

### Changes made after July 2019 Telecon Reading

naveen-rn commented on content/shmem_quiet.tex:
```diff
@@ -19,9 +19,9 @@
 \apidescription{
     The \FUNC{shmem\_quiet} routine ensures completion of \PUT{}, \ac{AMO},
-    memory store, and nonblocking \PUT{} and \GET{} routines on
+    \emph{put-with-signal}, memory store, and nonblocking \PUT{},\emph{put-with-signal}, and \GET{} routines on
```
The sh_q routine ensure completion of blocking put, amo, pws, and memory store, as well as nonblocking put, pws, and get routines on .....

and -> before memory store
and nonblocking -> as well as nonblocking

jdinan commented on content/shmem_fence.tex:
```diff
@@ -17,12 +19,18 @@
+    \ac{AMO}, and \emph{put-with-signal}, as well as nonblocking \PUT{},
+    and \emph{put-with-signal}
+    routines to symmetric data objects.  All memory store, blocking \PUT{},
+    \ac{AMO}, amd \emph{put-with-signal}, as well as nonblocking \PUT{},
```
```suggestion
    \ac{AMO}, and \emph{put-with-signal}, as well as nonblocking \PUT{},
```

davidozog replied:
```suggestion
    \ac{AMO}, amd \emph{put-with-signal}, as well as nonblocking \PUT{}
```

davidozog commented on content/shmem_fence.tex:
```diff
@@ -1,6 +1,8 @@
-    to symmetric data objects.
+    Assures ordering of delivery of memory store, blocking \PUT{}, 
+    \ac{AMO}, and \emph{put-with-signal}, as well as nonblocking 
+    \PUT{}, and \emph{put-with-signal} 
```
```suggestion
    \PUT{} and \emph{put-with-signal} 
```

davidozog commented on content/shmem_fence.tex:
```diff
@@ -17,12 +19,18 @@
     to the call to \FUNC{shmem\_fence} are guaranteed to be delivered before any
-    subsequent \PUT{}, \ac{AMO}, memory store, and nonblocking \PUT{} routines to symmetric data
+    subsequent memory store, blocking \PUT{}, \ac{AMO}, and \emph{put-with-signal}, 
+    as well as nonblocking \PUT{}, and \emph{put-with-signal}
```
```suggestion
    as well as nonblocking \PUT{} and \emph{put-with-signal}
```

davidozog commented on content/shmem_fence.tex:
```diff
@@ -42,17 +50,19 @@
     \FUNC{shmem\_fence} and \FUNC{shmem\_quiet}, in that, \FUNC{shmem\_quiet}
-    guarantees completion of \PUT{}, \ac{AMO}, memory store, and nonblocking \PUT{} routines to
+    guarantees completion of memory store, blocking \PUT{}, \ac{AMO}, and
+    \emph{put-with-signal}, as well as nonblocking \PUT{}, and \emph{put-with-signal} routines to
```
```suggestion
    \emph{put-with-signal}, as well as nonblocking \PUT{} and \emph{put-with-signal} routines to
```

davidozog commented on content/shmem_fence.tex:
```diff
@@ -42,17 +50,19 @@
-    \ac{AMO}, memory store, and nonblocking \PUT{} routines to symmetric data objects is desired
+    The \FUNC{shmem\_quiet} routine should be called if completion of memory
+    store, blocking \PUT{}, \ac{AMO}, and \emph{put-with-signal}, as well as
+    nonblocking \PUT{}, and \emph{put-with-signal} routines to symmetric data objects is desired
```
```suggestion
    nonblocking \PUT{} and \emph{put-with-signal} routines to symmetric data objects is desired
```
